src/core/ext/filters/deadline/deadline_filter.cc

```diff
@@ -69,12 +69,13 @@ static void timer_callback(void* arg, grpc_error* error) {
         GRPC_ERROR_CREATE_FROM_STATIC_STRING("Deadline Exceeded"),
         GRPC_ERROR_INT_GRPC_STATUS, GRPC_STATUS_DEADLINE_EXCEEDED);
     deadline_state->call_combiner->Cancel(GRPC_ERROR_REF(error));
-    GRPC_CLOSURE_INIT(&deadline_state->timer_callback,
-                      send_cancel_op_in_call_combiner, elem,
-                      grpc_schedule_on_exec_ctx);
-    GRPC_CALL_COMBINER_START(deadline_state->call_combiner,
-                             &deadline_state->timer_callback, error,
-                             "deadline exceeded -- sending cancel_stream op");
+    // The inlined closure might still be in use (holding a pending call), so
```

Comment:
In what sequence of events is this occurring?  The only path I can see where this could happen is `start_timer_if_needed()` creates a new timer since the original timer is pending, and that's still true by the time we get here.  Can we confirm that that's actually what's happening?  Do we know why this is being triggered now when it wasn't before, and why it's occurring only in this one xDS test?

Reply:
And if this is a problem, why are we not also running into the same problem when `deadline_timer->timer_callback` is used inside of `send_cancel_op_in_call_combiner()`?

Depending on what the problem really is (I'd really like to fully understand that before we agree on *any* fix), I think #24575 might be a more general approach.  Can one of you please try that out with the absl change and see if it fixes the problem?

Reply:
Yes, that is the path I saw too for this bug. Based on the logs at https://github.com/grpc/grpc/issues/24549, we do see that the race is occurring at the access of `deadline_state->closure` in this function `timer_callback()` while the closure is still stuck in the ExecCtx, and it looks like this is the only access outside the call combiner.

After patching this in, @donnadionne verified that the data race no longer exists. 

I was also verifying the other uses of `timer_callback` and the one inside `send_cancel_op_in_call_combiner()` also seemed doubtful, and it does seem to be the case that the initial closure can still be pending even when `send_cancel_op_in_call_combiner()` is invoked. Updated.

@donnadionne can you verify that this is indeed the sequence of events that we are running into?

Reply:
Also, I like the approach in #24575 too

Reply:
It's not clear to me that the logs in #24549 provide enough data to indicate exactly what's happening here.  They do show that the closure is in the `ExecCtx` queue, but they don't show how it got there.  How do we know this is from `start_timer_if_needed()` as opposed to some other code path?  For example, it could also be the case that the timer itself has fired and has invoked the callback already, but the `ExecCtx` has offloaded it to another thread for some reason.

To be clear, I'm not saying that I think that's likely; I actually do think that the `start_timer_if_needed()` code-path is the most likely culprit.  I'm just saying that I want to get proof instead of making an assumption.  Othewise we don't know for sure that we've really solved the problem.

Can we maybe add some logging to the deadline filter to verify the sequence of events here?

Reply:
I added logs here:
@@ -96,6 +115,7 @@ static void start_timer_if_needed(grpc_call_element* elem,
       // Note: We do not start the timer if there is already a timer
       return;
     case GRPC_DEADLINE_STATE_FINISHED:
+      gpr_log(GPR_INFO, "DONNAAA GRPC_DEADLINE_STATE_FINISHED");
       deadline_state->timer_state = GRPC_DEADLINE_STATE_PENDING;
       // If we've already created and destroyed a timer, we always create a
       // new closure: we have no other guarantee that the inlined closure is
@@ -104,12 +124,23 @@ static void start_timer_if_needed(grpc_call_element* elem,
           GRPC_CLOSURE_CREATE(timer_callback, elem, grpc_schedule_on_exec_ctx);
       break;
     case GRPC_DEADLINE_STATE_INITIAL:
+      gpr_log(GPR_INFO, "DONNAAA GRPC_DEADLINE_STATE_INITIAL");
       deadline_state->timer_state = GRPC_DEADLINE_STATE_PENDING;
       closure =
           GRPC_CLOSURE_INIT(&deadline_state->timer_callback, timer_callback,
                             elem, grpc_schedule_on_exec_ctx);
       break;
   }
+      void *stack[50];
+      int size = absl::GetStackTrace(stack, 48, 2);
+      for (int i = 0; i < size; ++i) {
+        char out[256];
+        if (absl::Symbolize(stack[i], out, 256)) {
+          gpr_log(GPR_INFO, "donna stack trace:%s and %s", stack[i], out);
+        } else {
+          gpr_log(GPR_INFO, "donna stack trace:%s", stack[i]);
+        }
+      }

Please see test logs here:
https://source.cloud.google.com/results/invocations/27839f78-5173-432f-bd75-dbd80de666d0/targets/%2F%2Ftest%2Fcpp%2Fend2end:xds_end2end_test@poller%3Depoll1/log

I know we went into both
case GRPC_DEADLINE_STATE_FINISHED: and
case GRPC_DEADLINE_STATE_INITIAL:

Reply:
I also tested and #24575 also seems to solve the issue

Reply:
Thanks, Donna.  That log does seem to confirm that this is the problem.

Okay, in that case, let's go with #24575, since it seems to be a cleaner approach.

Reply:
Thanks for verifying Donna! Yes, let's go with #24575 
 
(As an aside, the closures that get offloaded are the combiner closures, not the ExecCtx closures, but it's still good to verify the codepath.)